src/ncp/ncp_spinel.cpp

```diff
@@ -181,6 +183,7 @@ void NcpSpinel::HandleResponse(spinel_tid_t aTid, const uint8_t *aFrame, uint16_
         SuccessOrExit(error = SpinelDataUnpack(data, len, SPINEL_DATATYPE_UINT_PACKED_S, &spinelRole));
 
         mDeviceRole = SpinelRoleToDeviceRole(spinelRole);
+        mPropsObserver->SetDeviceRole(mDeviceRole);
```

Comment:
How do you guarantee that the device role is always propagated back to the host before the callback of `registerStateCallback()` is invoked?

Reply:
Because the state callback is triggered by the update of the properties.

For example, the device role is updated on NCP -> NCP sends a notification to the host about the change -> the state callback is invoked on host.